File: AI/graph/town_graph.py
"""
LangGraph Integration - Town Orchestrator
小镇总编排图 - 管理多个智能体的协作
"""

from typing import TypedDict, List, Dict, Optional, Any
from langchain_core.messages import BaseMessage, HumanMessage, AIMessage
import asyncio

from AI.graph.agent_graph import AgentGraph, AgentState
from AI.agents.base_agent import BaseAgent
from AI.memory.dual_memory import DualMemorySystem, create_memory_system
from AI.skills.skill_system import SkillRegistry, get_skill_registry
from AI.emotion.emotion_engine import EmotionEngine, get_emotion_engine


class TownState(TypedDict):
    """
    小镇状态
    
    Attributes:
        user_input: 用户输入
        target_agent_id: 目标智能体 ID
        selected_agent_id: 实际选中的智能体 ID
        agent_response: 智能体回复
        should_continue: 是否继续
    """
    user_input: str
    target_agent_id: str
    selected_agent_id: str
    agent_response: str
    conversation_history: List[BaseMessage]
    should_continue: bool


class TownOrchestrator:
    """
    小镇编排器
    
    基于 LangGraph 构建的多智能体协作系统
    """
    
    def __init__(
        self,
        town_name: str = "赛博小镇",
    ):
        """
        初始化小镇编排器
        
        Args:
            town_name: 小镇名称
        """
        self.town_name = town_name
        
        # 智能体注册表
        self._agents: Dict[str, BaseAgent] = {}
        
        # 每个智能体的图和记忆系统
        self._agent_graphs: Dict[str, AgentGraph] = {}
        self._agent_memories: Dict[str, DualMemorySystem] = {}
        
        # 共享组件
        self._skill_registry: SkillRegistry = get_skill_registry()
        self._emotion_engine: EmotionEngine = get_emotion_engine()
        
        # 当前状态
        self.state: TownState = {
            "user_input": "",
            "target_agent_id": "",
            "selected_agent_id": "",
            "agent_response": "",
            "conversation_history": [],
            "should_continue": True,
        }
    
    def add_agent(
        self,
        agent: BaseAgent,
        memory_window: int = 5,
    ) -> None:
        """
        添加智能体到小镇
        
        Args:
            agent: 智能体实例
            memory_window: 记忆窗口大小
        """
        self._agents[agent.agent_id] = agent
        
        # 创建记忆系统
        memory = create_memory_system(
            agent_id=agent.agent_id,
            window_size=memory_window,
        )
        self._agent_memories[agent.agent_id] = memory
        
        # 注意：AgentGraph 需要 other_agent_id，这里先不创建
        # 在第一次对话时动态创建
    
    def remove_agent(self, agent_id: str) -> bool:
        """移除智能体"""
        if agent_id in self._agents:
            del self._agents[agent_id]
            
            if agent_id in self._agent_graphs:
                del self._agent_graphs[agent_id]
            
            if agent_id in self._agent_memories:
                del self._agent_memories[agent_id]
            
            return True
        return False
    
    def get_agent(self, agent_id: str) -> Optional[BaseAgent]:
        """获取智能体"""
        return self._agents.get(agent_id)
    
    def list_agents(self) -> List[Dict]:
        """列出所有智能体"""
        return [agent.get_status() for agent in self._agents.values()]
    
    def _get_or_create_agent_graph(
        self,
        agent_id: str,
        other_agent_id: str,
    ) -> Optional[AgentGraph]:
        """获取或创建智能体图"""
        # 如果已存在，直接返回
        if agent_id in self._agent_graphs:
            return self._agent_graphs[agent_id]
        
        # 检查智能体是否存在
        if agent_id not in self._agents:
            return None
        
        agent = self._agents[agent_id]
        memory = self._agent_memories[agent_id]
        
        # 创建图
        graph = AgentGraph(
            agent=agent,
            memory_system=memory,
            skill_registry=self._skill_registry,
            emotion_engine=self._emotion_engine,
            other_agent_id=other_agent_id,
        )
        
        self._agent_graphs[agent_id] = graph
        return graph
    
    async def dispatch_node(self, state: TownState) -> TownState:
        """
        分发节点 - 根据用户输入选择目标智能体
        
        简单实现：直接使用用户指定的 target_agent_id
        进阶实现：可以用 LLM 分析意图，自动选择合适的智能体
        """
        if not state["target_agent_id"]:
            # 如果没有指定，选择第一个智能体
            if self._agents:
                state["selected_agent_id"] = list(self._agents.keys())[0]
            else:
                state["selected_agent_id"] = ""
        else:
            state["selected_agent_id"] = state["target_agent_id"]
        
        return state
    
    async def route_to_agent_node(self, state: TownState) -> TownState:
        """路由到智能体节点 - 调用对应智能体的图"""
        agent_id = state["selected_agent_id"]
        
        if not agent_id:
            state["agent_response"] = "[系统] 小镇还没有居民..."
            return state
        
        # 获取或创建智能体图
        # 使用用户作为对话方
        graph = self._get_or_create_agent_graph(agent_id, "user")
        
        if not graph:
            state["agent_response"] = f"[系统] 找不到居民 {agent_id}"
            return state
        
        # 处理消息
        result = await graph.process_message(
            user_input=state["user_input"],
            conversation_history=state["conversation_history"],
        )
        
        state["agent_response"] = result["response"]
        
        # 更新对话历史
        state["conversation_history"].append(HumanMessage(content=state["user_input"]))
        state["conversation_history"].append(AIMessage(content=result["response"]))
        
        return state
    
    def should_continue(self, state: TownState) -> str:
        """判断是否继续对话"""
        if state["should_continue"]:
            return "continue"
        return "end"
    
    async def chat(
        self,
        user_input: str,
        target_agent_id: Optional[str] = None,
        conversation_history: Optional[List[BaseMessage]] = None,
    ) -> Dict:
        """
        与小镇居民聊天
        
        Args:
            user_input: 用户输入
            target_agent_id: 目标智能体 ID（可选）
            conversation_history: 对话历史
            
        Returns:
            包含回复和状态信息的字典
        """
        # 初始化状态
        self.state = {
            "user_input": user_input,
            "target_agent_id": target_agent_id or "",
            "selected_agent_id": "",
            "agent_response": "",
            "conversation_history": conversation_history or [],
            "should_continue": True,
        }
        
        state = self.state
        
        # 1. 分发到合适的智能体
        state = await self.dispatch_node(state)
        
        # 2. 路由到智能体进行处理
        state = await self.route_to_agent_node(state)
        
        self.state = state
        
        return {
            "response": state["agent_response"],
            "agent_id": state["selected_agent_id"],
            "agent_name": self._agents[state["selected_agent_id"]].profile.name if state["selected_agent_id"] else "",
            "conversation_history": state["conversation_history"],
        }
    
    async def multi_agent_chat(
        self,
        topic: str,
        participant_ids: List[str],
        max_rounds: int = 3,
    ) -> List[Dict]:
        """
        多智能体讨论模式
        
        Args:
            topic: 讨论话题
            participant_ids: 参与讨论的智能体 ID 列表
            max_rounds: 最大轮数
            
        Returns:
            对话记录列表
        """
        if len(participant_ids) < 2:
            return [{"error": "至少需要两个智能体"}]
        
        records = []
        current_topic = topic
        
        for round_num in range(max_rounds):
            for i, agent_id in enumerate(participant_ids):
                # 确定上一个发言者
                prev_agent_id = participant_ids[i - 1] if i > 0 else "user"
                
                # 获取智能体图
                graph = self._get_or_create_agent_graph(agent_id, prev_agent_id)
                
                if not graph:
                    continue
                
                # 处理
                result = await graph.process_message(
                    user_input=current_topic,
                    conversation_history=[],
                )
                
                records.append({
                    "round": round_num + 1,
                    "agent_id": agent_id,
                    "agent_name": self._agents[agent_id].profile.name,
                    "response": result["response"],
                })
                
                # 更新话题为上一个回复
                current_topic = result["response"]
        
        return records
    
    def get_town_status(self) -> Dict:
        """获取小镇状态"""
        return {
            "town_name": self.town_name,
            "agent_count": len(self._agents),
            "agents": self.list_agents(),
            "skill_count": len(self._skill_registry.get_all_skills()),
        }
    
    async def cleanup(self) -> None:
        """清理资源"""
        # 清理所有技能
        await self._skill_registry.cleanup_all()
        
        # 清理所有 MCP
        from mcp.lazy_loader import get_mcp_loader
        loader = get_mcp_loader()
        await loader.unload_all()


def create_town_orchestrator(
    town_name: str = "赛博小镇",
    agents: Optional[List[BaseAgent]] = None,
) -> TownOrchestrator:
    """
    工厂函数：创建小镇编排器
    
    Args:
        town_name: 小镇名称
        agents: 智能体列表
        
    Returns:
        TownOrchestrator 实例
    """
    orchestrator = TownOrchestrator(town_name=town_name)
    
    if agents:
        for agent in agents:
            orchestrator.add_agent(agent)
    
    return orchestrator


# 示例：创建默认小镇
async def create_default_town(num_agents: int = 4) -> TownOrchestrator:
    """
    创建默认小镇
    
    Args:
        num_agents: 智能体数量
        
    Returns:
        TownOrchestrator 实例
    """
    from AI.agents.models import DEFAULT_PROFILES, Profession
    from AI.agents.base_agent import BaseAgent
    
    orchestrator = TownOrchestrator(town_name="赛博小镇")
    
    # 获取LLM模型
    from config import get_llm
    llm = get_llm()
    
    # 添加默认智能体
    profiles = list(DEFAULT_PROFILES.values())[:num_agents]
    
    for i, profile in enumerate(profiles):
        agent = BaseAgent(
            agent_id=f"agent_{i}",
            profile=profile,
            llm_model=llm,
        )
        orchestrator.add_agent(agent)
    
    return orchestrator

File: src/AI/graph/town_graph.ts
/**
 * 赛博小镇 - 小镇图谱系统
 * 
 * 这个文件是整个小镇的核心管理系统，负责：
 * 1. 管理所有角色（智能体）的创建、添加、删除
 * 2. 处理用户与角色的对话流程
 * 3. 支持多角色之间的群聊
 * 4. 集成记忆系统、技能系统、情感系统
 * 
 * 就像是一个小镇的市长办公室：
 * - 管理所有居民（角色）
 * - 安排居民与访客（用户）的会面
 * - 组织居民之间的交流活动
 * - 记录每个居民的记忆和技能
 */

// 导入智能体相关类和函数
// BaseMessage: 消息类型定义
// HumanMessage: 人类消息
// AIMessage: AI消息
// BaseAgent: 基础智能体类
// createBaseAgent: 创建智能体的工厂函数
import { BaseMessage, HumanMessage, AIMessage, BaseAgent, createBaseAgent } from '../agents/base_agent';
// AgentGraph: 智能体图谱，处理单个智能体的对话逻辑
import { AgentGraph } from './agent_graph';
// DualMemorySystem: 双重记忆系统（短期记忆+长期记忆）
import { DualMemorySystem, createMemorySystem } from '../memory/dual_memory';
// DEFAULT_PROFILES: 8个默认角色的配置
import { DEFAULT_PROFILES } from '../agents/models';
// getSkillRegistry: 获取技能注册表
import { getSkillRegistry as getRealSkillRegistry } from '../skills/skill_system';

/**
 * 小镇状态接口，定义了小镇的当前状态
 * 
 * 这个接口描述了小镇在处理用户请求时的所有相关信息。
 * 就像是一个表单，记录了：
 * - 用户说了什么（user_input）
 * - 用户想和谁对话（target_agent_id）
 * - 实际选择了哪个角色（selected_agent_id）
 * - 角色的回复是什么（agent_response）
 * - 完整的对话历史（conversation_history）
 */
export interface TownState {
  user_input: string;           // 用户输入的消息内容
  target_agent_id: string;      // 用户想要对话的目标角色ID
  selected_agent_id: string;    // 实际选中的角色ID（可能与目标不同）
  agent_response: string;       // 角色的回复内容
  conversation_history: BaseMessage[]; // 完整的对话历史记录
  should_continue: boolean;     // 是否继续对话流程（true=继续，false=结束）
}

/**
 * 技能注册表接口，定义了技能管理的方法
 */
export interface SkillRegistry {
  get_all_skills: () => any[];            // 获取所有技能
  cleanup_all: () => Promise<void>;       // 清理所有技能
  findMatchingSkills: (input: string) => any[];  // 查找匹配的技能
  getSkill: (name: string) => any;        // 获取技能
}

/**
 * 情感引擎接口，定义了情感管理的方法
 */
export interface EmotionEngine {
  getRelationshipInfo: (agentId1: string, agentId2: string) => any;  // 获取关系信息
  getConversationStyleHint: (level: any) => string;                 // 获取对话风格提示
  interact: (params: any) => any;                                  // 互动
}

/**
 * 获取技能注册表
 * @returns 技能注册表实例
 */
export function getSkillRegistry(): SkillRegistry {
  // 使用真实的技能注册表
  const realRegistry = getRealSkillRegistry();
  
  return {
    get_all_skills: () => realRegistry.getAllSkills(),
    cleanup_all: () => realRegistry.cleanupAll(),
    findMatchingSkills: (input: string) => realRegistry.findMatchingSkills(input),
    getSkill: (name: string) => realRegistry.getSkill(name)
  };
}

/**
 * 获取情感引擎
 * @returns 情感引擎实例
 */
export function getEmotionEngine(): EmotionEngine {
  return {
    getRelationshipInfo: (agentId1: string, agentId2: string) => null,
    getConversationStyleHint: (level: any) => "",
    interact: (params: any) => { return { level_changed: false }; }
  };
}

/**
 * 小镇编排器类，管理小镇中的智能体和交互
 * 
 * 这是整个小镇的核心管理类，负责协调所有角色和系统。
 * 
 * 主要功能：
 * 1. 添加/删除角色（addAgent/removeAgent）
 * 2. 处理用户与角色的对话（chat方法）
 * 3. 支持多角色群聊（multiAgentChat方法）
 * 4. 管理每个角色的记忆、技能、情感系统
 * 
 * 就像是一个小镇的管理中心：
 * - 有居民名单（agents）
 * - 有每个居民的档案（agentGraphs）
 * - 有居民的记忆库（agentMemories）
 * - 有技能培训中心（skillRegistry）
 * - 有情感咨询师（emotionEngine）
 */
export class TownOrchestrator {
  private townName: string;                // 小镇名称，比如“赛博小镇”
  private agents: Record<string, BaseAgent>;       // 所有角色的集合，键是角色ID，值是角色对象
  private agentGraphs: Record<string, AgentGraph>;  // 每个角色的图谱，用于处理对话逻辑
  private agentMemories: Record<string, DualMemorySystem>; // 每个角色的记忆系统
  private skillRegistry: SkillRegistry;     // 技能注册表，管理所有可用技能
  private emotionEngine: EmotionEngine;     // 情感引擎，管理角色间的情感关系
  private state: TownState;                // 小镇的当前状态

  /**
   * 获取小镇名称
   */
  get TownName(): string {
    return this.townName;
  }

  /**
   * 获取智能体集合
   */
  get Agents(): Record<string, BaseAgent> {
    return this.agents;
  }

  /**
   * 构造函数 - 创建小镇编排器实例
   * 
   * 就像是建立一个新的小镇管理中心：
   * 1. 给小镇起个名字（默认“赛博小镇”）
   * 2. 初始化居民名单（空列表）
   * 3. 准备档案室（agentGraphs）
   * 4. 准备记忆库（agentMemories）
   * 5. 连接技能中心（skillRegistry）
   * 6. 连接情感咨询室（emotionEngine）
   * 7. 设置初始状态
   * 
   * @param townName 小镇名称，默认为“赛博小镇”
   */
  constructor(townName: string = "赛博小镇") {
    this.townName = townName;  // 保存小镇名称
    this.agents = {};  // 初始化空的角色集合
    this.agentGraphs = {};  // 初始化空的图谱集合
    this.agentMemories = {};  // 初始化空的记忆系统集合
    this.skillRegistry = getSkillRegistry();  // 获取技能注册表实例
    this.emotionEngine = getEmotionEngine();  // 获取情感引擎实例
  
    // 初始化小镇状态
    this.state = {
      user_input: "",  // 初始用户输入为空
      target_agent_id: "",  // 初始目标角色为空
      selected_agent_id: "",  // 初始选中角色为空
      agent_response: "",  // 初始回复为空
      conversation_history: [],  // 初始对话历史为空数组
      should_continue: true  // 默认继续处理
    };
  }

  /**
   * 添加智能体 - 向小镇添加一个新角色
   * 
   * 就像是为小镇招募一个新居民：
   * 1. 把角色加入居民名单（agents）
   * 2. 为角色创建记忆系统（用于记住对话）
   * 3. 设置记忆窗口大小（默认保留最近5条对话）
   * 
   * @param agent 要添加的角色实例
   * @param memoryWindow 记忆窗口大小，默认5，表示保留最近5条对话在短期记忆中
   */
  addAgent(agent: BaseAgent, memoryWindow: number = 5): void {
    // 将角色添加到居民名单中，键是角色ID
    this.agents[agent.AgentId] = agent;

    // 为角色创建记忆系统
    // 记忆系统包括短期记忆（最近对话）和长期记忆（向量数据库）
    const memory = createMemorySystem(
      agent.AgentId,  // 角色ID，用于区分不同角色的记忆
      memoryWindow    // 记忆窗口大小
    );
    // 保存角色的记忆系统
    this.agentMemories[agent.AgentId] = memory;
  }

  /**
   * 移除智能体
   * @param agentId 智能体ID
   * @returns 是否成功移除
   */
  removeAgent(agentId: string): boolean {
    if (agentId in this.agents) {
      delete this.agents[agentId];

      // 清理相关资源
      if (agentId in this.agentGraphs) {
        delete this.agentGraphs[agentId];
      }

      if (agentId in this.agentMemories) {
        delete this.agentMemories[agentId];
      }

      return true;
    }
    return false;
  }

  /**
   * 获取智能体
   * @param agentId 智能体ID
   * @returns 智能体实例或undefined
   */
  getAgent(agentId: string): BaseAgent | undefined {
    return this.agents[agentId];
  }

  /**
   * 列出所有智能体
   * @returns 智能体状态数组
   */
  listAgents(): Record<string, any>[] {
    return Object.values(this.agents).map(agent => agent.getStatus());
  }

  /**
   * 获取或创建智能体图
   * @param agentId 智能体ID
   * @param otherAgentId 其他智能体ID
   * @returns 智能体图实例或null
   */
  private getOrCreateAgentGraph(agentId: string, otherAgentId: string): AgentGraph | null {
    if (agentId in this.agentGraphs) {
      return this.agentGraphs[agentId];
    }

    if (!(agentId in this.agents)) {
      return null;
    }

    const agent = this.agents[agentId];
    const memory = this.agentMemories[agentId];

    // 创建智能体图
    const graph = new AgentGraph(
      agent,
      memory,
      this.skillRegistry as any,
      this.emotionEngine as any,
      otherAgentId
    );

    this.agentGraphs[agentId] = graph;
    return graph;
  }

  /**
   * 分发节点 - 确定要和哪个角色对话
   * 
   * 这个方法的职责很简单：
   * 1. 如果用户指定了目标角色（target_agent_id不为空），就使用指定的角色
   * 2. 如果用户没有指定角色（target_agent_id为空），就选择第一个可用角色
   * 
   * @param state 小镇状态，包含用户输入和目标角色ID
   * @returns 更新后的状态，包含选中的角色ID
   */
  async dispatchNode(state: TownState): Promise<TownState> {
    // 情况1：用户明确指定了目标角色
    if (state.target_agent_id) {
      // 直接使用用户指定的角色ID
      state.selected_agent_id = state.target_agent_id;
    } 
    // 情况2：用户没有指定角色
    else {
      // 获取所有可用的角色ID列表
      const agentIds = Object.keys(this.agents);
      
      // 如果有可用角色，选择第一个；否则设为空字符串
      if (agentIds.length > 0) {
        state.selected_agent_id = agentIds[0];
      } else {
        state.selected_agent_id = "";
      }
    }

    return state;
  }

  /**
   * 路由到智能体节点
   * @param state 小镇状态
   * @returns 更新后的状态
   */
  async routeToAgentNode(state: TownState): Promise<TownState> {
    const agentId = state.selected_agent_id;

    if (!agentId) {
      state.agent_response = "[系统] 小镇还没有居民...";
      return state;
    }

    const graph = this.getOrCreateAgentGraph(agentId, "user");

    if (!graph) {
      state.agent_response = `[系统] 找不到居民 ${agentId}`;
      return state;
    }

    // 处理消息
    const result = await graph.processMessage(
      state.user_input,
      state.conversation_history
    );

    state.agent_response = result.response;

    // 更新对话历史
    state.conversation_history.push(new HumanMessage(state.user_input));
    state.conversation_history.push(new AIMessage(result.response));

    return state;
  }


  /**
   * 聊天 - 处理用户与角色的对话
   * 
   * 这是小镇最核心的功能，当用户发送消息时，会调用这个方法。
   * 
   * 工作流程：
   * 1. 初始化小镇状态（记录用户输入、目标角色等）
   * 2. 分发节点（dispatchNode）- 确定要和哪个角色对话
   * 3. 路由到角色节点（routeToAgentNode）- 调用角色的respond方法生成回复
   * 4. 返回回复给用户
   * 
   * 就像是一个接待员：
   * - 接收访客的问题
   * - 确定应该由哪位居民回答
   * - 安排会面并获取回答
   * - 把回答转告给访客
   * 
   * @param userInput 用户输入的消息，比如“你好”
   * @param targetAgentId 目标角色ID（可选），如果不指定，会自动选择第一个角色
   * @param conversationHistory 对话历史（可选），包含之前的对话记录
   * @returns 包含回复内容、角色ID、角色名、对话历史的对象
   */
  async chat(
    userInput: string,
    targetAgentId?: string,
    conversationHistory?: BaseMessage[]
  ): Promise<Record<string, any>> {
    // 初始化小镇状态
    // 就像是填写一个接待表单
    this.state = {
      user_input: userInput,  // 记录用户说的话
      target_agent_id: targetAgentId || "",  // 记录用户想和谁对话（如果没有则为空）
      selected_agent_id: "",  // 实际选中的角色（稍后确定）
      agent_response: "",  // 角色的回复（稍后填充）
      conversation_history: conversationHistory || [],  // 对话历史（如果没有则为空数组）
      should_continue: true  // 默认继续处理流程
    };

    let state = this.state;  // 获取当前状态

    // 第一步：分发节点 - 确定要和哪个角色对话
    // 如果用户指定了目标角色，就使用指定的；否则自动选择第一个角色
    state = await this.dispatchNode(state);
    
    // 第二步：路由到角色节点 - 调用角色的respond方法生成回复
    // 这里会加载角色的记忆、技能，然后调用AI模型生成回复
    state = await this.routeToAgentNode(state);

    // 更新小镇状态
    this.state = state;

    // 返回结果
    return {
      response: state.agent_response,  // 角色的回复内容
      agent_id: state.selected_agent_id,  // 实际对话的角色ID
      agent_name: this.agents[state.selected_agent_id]?.Profile.name || "",  // 角色名字
      conversation_history: state.conversation_history  // 更新后的对话历史
    };
  }

  /**
   * 多智能体聊天
   * @param topic 话题
   * @param participantIds 参与者ID数组
   * @param maxRounds 最大轮数（默认3）
   * @returns 聊天记录
   */
  async multiAgentChat(
    topic: string,
    participantIds: string[],
    maxRounds: number = 3
  ): Promise<Record<string, any>[]> {
    if (participantIds.length < 2) {
      return [{ error: "至少需要两个智能体" }];
    }

    const records: Record<string, any>[] = [];
    let currentTopic = topic;

    // 多轮对话
    for (let roundNum = 0; roundNum < maxRounds; roundNum++) {
      for (let i = 0; i < participantIds.length; i++) {
        const agentId = participantIds[i];
        const prevAgentId = i > 0 ? participantIds[i - 1] : "user";

        const graph = this.getOrCreateAgentGraph(agentId, prevAgentId);

        if (!graph) {
          continue;
        }

        // 处理消息
        const result = await graph.processMessage(
          currentTopic,
          []
        );

        // 记录结果
        records.push({
          round: roundNum + 1,
          agent_id: agentId,
          agent_name: this.agents[agentId]?.Profile.name || "",
          response: result.response
        });

        // 更新话题
        currentTopic = result.response;
      }
    }

    return records;
  }

  /**
   * 获取小镇状态
   * @returns 小镇状态对象
   */
  getTownStatus(): Record<string, any> {
    return {
      town_name: this.townName,
      agent_count: Object.keys(this.agents).length,
      agents: this.listAgents(),
      skill_count: this.skillRegistry.get_all_skills().length
    };
  }

  /**
   * 清理资源
   */
  async cleanup(): Promise<void> {
    await this.skillRegistry.cleanup_all();

    // 清理所有 MCP
    // 暂时跳过，后续实现
  }
}

/**
 * 创建小镇编排器
 * @param townName 小镇名称（默认"赛博小镇"）
 * @param agents 智能体数组（可选）
 * @returns 小镇编排器实例
 */
export function createTownOrchestrator(
  townName: string = "赛博小镇",
  agents?: BaseAgent[]
): TownOrchestrator {
  const orchestrator = new TownOrchestrator(townName);

  if (agents) {
    for (const agent of agents) {
      orchestrator.addAgent(agent);
    }
  }

  return orchestrator;
}

/**
 * 创建默认小镇
 * @param numAgents 智能体数量（默认4）
 * @returns 小镇编排器实例
 */
export async function createDefaultTown(numAgents: number = 4): Promise<TownOrchestrator> {
  const orchestrator = new TownOrchestrator("赛博小镇");

  // 使用默认配置创建智能体
  const profileKeys = Object.keys(DEFAULT_PROFILES);
  const selectedProfiles = profileKeys.slice(0, numAgents);

  for (const key of selectedProfiles) {
    const profile = DEFAULT_PROFILES[key];
    const agent = createBaseAgent(
      profile.name,
      profile.profession,
      `agent_${key}`,
      {
        age: profile.age,
        personality: profile.personality,
        background: profile.background
      }
    );
    // 手动设置爱好和说话风格
    (agent as any).profile.hobbies = profile.hobbies;
    (agent as any).profile.speech_style = profile.speech_style;
    (agent as any).profile.appearance = profile.appearance;
    orchestrator.addAgent(agent);
  }

  return orchestrator;
}
